canopen/pdo/base.py

```diff
@@ -78,8 +78,16 @@ def subscribe(self):
     def export(self, filename):
         """Export current configuration to a database file.
 
+        .. note::
+           This API depends on the :mod:`!canmatrix` optional dependency,
+           which can be installed by requesting the ``db_export`` feature::
+
+              python3 -m pip install 'canopen[db_export]'
+
         :param str filename:
             Filename to save to (e.g. DBC, DBF, ARXML, KCD etc)
+        :raises ImportError:
+            When the ``db_export`` feature is not installed.
```

Comment:
Moving the imports to the top is certainly a good idea.  I suggest to try the import at the top, but pass (with a comment) when it fails.  Down here in the method, let's be pythonic and ask for forgiveness rather than permission:  Just `try` the first line and if the import was not successful, catch the `NameError` and re-raise an appropriate `RuntimeError` or further specialized exception explaining what needs to be done.

I'd like to avoid any special constructs such as sentinel variables or dummy values for the `canmatrix` name.

Reply:
Sure, that's a fine approach!

Reply:
I choose NotImplementedError. See 013ff93cacb4261dd9b5c95dd7bb139056abab00. Let me know if that's fine.

Reply:
Thought we wanted to use import at the top though?  I meant wrapping the actual `CanMatrix` construction in the `try... except`.

Reply:
> Thought we wanted to use import at the top though?

Why?

Reply:
To expand on my "why?" comment: Doing an `.export()` with `canmatrix` installed works; doing it without `canmatrix` installed raises `NotImplementedError` with an explanatory error message. What's the missing behaviour?

Reply:
My understanding was that local imports are frowned upon.  Should we even introduce additional functionality based on `canmatrix`, then doing it once at the top seems the better approach.  So why not start with it right away?

Basically, have all dependencies indicated at the top, even the optional one (with an appropriate comment why `pass`ing on the failed import is OK).  In the method, as the first thing to do, check explicitly if the required optional dependency is there (by catching `NameError`), so in that place it's also easily spotted that the option is a prerequisite.

Reply:
> My understanding was that local imports are frowned upon.

IMO, that depends on the API in question, and the kind of code base. For tests, I see it all the time. For optional APIs like this, I think it is a valid use case.

> Should we even introduce additional functionality based on `canmatrix`, then doing it once at the top seems the better approach. So why not start with it right away?

I'm not convinced by that argument, but I'll abide :)

Reply:
FTR; I won't be back on my computer until earliest tomorrow, so don't expect any immediate updates. In the mean time, consider if you really want to go this route; I'm not sure the added complexity is really worth it. The current solution is compact and accurate. I'm not sure the proposed alternative is a better solution.

Reply:
See https://github.com/acolomb/canopen/tree/fix-pdo-export for what I had in mind.  Passes tests and I don't get any related mypy errors either.

Reply:
> See https://github.com/acolomb/canopen/tree/fix-pdo-export for what I had in mind. Passes tests and I don't get any related mypy errors either.

Sure, I knew what you meant, but thanks; I just think 013ff93cacb4261dd9b5c95dd7bb139056abab00 is a nicer approach 😁

Reply:
If my opinion carries any weight I'd vote for the latter proposal by @erlend-aasland . I think its compact and to the point. and the simplest of the two. The downside to it is that linters will complain about import inside functions and have to be muted. Conditional run-time dependencies requires some special trickeries regardless how its solved.

Reply:
> The downside to it is that linters will complain about import inside functions and have to be muted. Conditional run-time dependencies requires some special trickeries regardless how its solved.

This PR does not add new imports; not at module level, nor from within methods; **the imports are already there**.

You linter can be configured to not complain about imports from within a function/method.